tests/php/Forms/RequiredFieldsTest.php

```diff
@@ -322,4 +324,71 @@ public function testHasOneRelationFieldInterfaceValidation(string $className)
         // '1' passes required field validation
         $this->assertTrue($validator->php(['TestField' => '1']));
     }
+
+    public static function provideAllowWhitespaceOnly(): array
+    {
+        return [
+            'no-ws' => [
+                'value' => 'abc',
+                'expectedTrue' => true,
+                'expectedFalse' => true,
+            ],
+            'left-ws' => [
+                'value' => 'abc',
+                'expectedTrue' => true,
+                'expectedFalse' => true,
+            ],
+            'right-ws' => [
+                'value' => 'abc',
+                'expectedTrue' => true,
+                'expectedFalse' => true,
+            ],
+            'both-ws' => [
+                'value' => 'abc',
+                'expectedTrue' => true,
+                'expectedFalse' => true,
+            ],
+            'only-ws' => [
+                'value' => ' ',
+                'expectedTrue' => true,
+                'expectedFalse' => false,
+            ],
+            'only-ws-nbsp' => [
+                'value' => "\xc2\xa0",
+                'expectedTrue' => true,
+                'expectedFalse' => false,
+            ],
+            'only-ws-unicode' => [
+                // zero width no-break space
+                'value' => "\u{2028}",
+                'expectedTrue' => true,
+                'expectedFalse' => false,
+            ],
+            'no-value' => [
+                'value' => '',
+                'expectedTrue' => false,
+                'expectedFalse' => false,
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider provideAllowWhitespaceOnly
+     */
+    public function testAllowWhitespaceOnly(string $value, bool $expectedTrue, bool $expectedFalse): void
```

Comment:
It's not clear what `expectedTrue` and `expectedFalse` mean.... I _think_ it's this:

- `expectedTrue` means "if whitespace is _not_ allowed, and this boolean is `true`, the value should be valid
- `expectedFalse` means "if whitespace _is_ allowed, and this boolean is `true`, the value should be valid

If that's the case, then one of these two things would make it a lot clearer:

1. Have `allowWhiteSpace` and `expected`
  - This would mean having a `allowWhiteSpace` true and `allowWhiteSpace` false scenario for each value
  - It's REALLY clear what's happening
  - If one fails, the other is still tested (right now if the first fails the second won't be tested)
1. Have `validIfAllowedWhitespace` and `validIfNotAllowedWhitespace`
  - This allows you to keep the test as it is and only rename these params
  - It's clearer than it currently is, though not as clear as 1. above
  - If the first fails in your `foreach` loop, the second never gets tested


I prefer 1 above, but I'll accept either.
If my understanding of those params is wrong, then please use your best judgment to rename them to something clear, or add PHPDocs for them, or something.

Reply:
Updated